// skybridge/dist/commands/build.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { cpSync, rmSync } from "node:fs";
import { Command } from "@oclif/core";
import { Box, render, Text } from "ink";
import { useEffect } from "react";
import { Header } from "../cli/header.js";
import { useExecuteSteps } from "../cli/use-execute-steps.js";
export const commandSteps = [
    {
        label: "Building widgets",
        command: "vite build -c web/vite.config.ts",
    },
    {
        label: "Compiling server",
        run: () => rmSync("dist", { recursive: true, force: true }),
        command: "tsc -p tsconfig.server.json",
    },
    {
        label: "Copying static assets",
        run: () => cpSync("web/dist", "dist/assets", { recursive: true }),
    },
];
export default class Build extends Command {
    static description = "Build the widgets and MCP server";
    static examples = ["skybridge build"];
    static flags = {};
    async run() {
        const App = () => {
            const { currentStep, status, error, execute } = useExecuteSteps(commandSteps);
            useEffect(() => {
                execute();
            }, [execute]);
            return (_jsxs(Box, { flexDirection: "column", padding: 1, children: [_jsx(Header, { version: this.config.version, children: _jsx(Text, { color: "green", children: " \u2192 building for production\u2026" }) }), commandSteps.map((step, index) => {
                        const isCurrent = index === currentStep && status === "running";
                        const isCompleted = index < currentStep || status === "success";
                        const isError = status === "error" && index === currentStep;
                        return (_jsx(Box, { marginBottom: 0, children: _jsxs(Text, { color: isError ? "red" : isCompleted ? "green" : "grey", children: [isError ? "✗" : isCompleted ? "✓" : isCurrent ? "⟳" : "○", " ", step.label] }) }, step.label));
                    }), status === "success" && (_jsx(Box, { marginTop: 1, children: _jsx(Text, { color: "green", bold: true, children: "\u2713 Build completed successfully!" }) })), status === "error" && error && (_jsxs(Box, { marginTop: 1, flexDirection: "column", children: [_jsx(Text, { color: "red", bold: true, children: "\u2717 Build failed" }), _jsx(Box, { marginTop: 1, flexDirection: "column", children: error.split("\n").map((line) => (_jsx(Text, { color: "red", children: line }, line))) })] }))] }));
        };
        render(_jsx(App, {}), {
            exitOnCtrlC: true,
            patchConsole: false,
        });
    }
}
