// skybridge/dist/web/components/modal-provider.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useSyncExternalStore } from "react";
import { McpAppAdaptor } from "../bridges";
const modalStyles = `
.sb-modal-backdrop {
  position: fixed;
  inset: 0;
  background: rgba(0, 0, 0, 0.5);
  z-index: 9998;
}
.sb-modal-container {
  border-radius: 12px;
  position: fixed;
  inset: 0;
  margin: auto;
  width: fit-content;
  height: fit-content;
  background: white;
  z-index: 9999;
}
`;
export function ModalProvider({ children }) {
    const adaptor = McpAppAdaptor.getInstance();
    const { mode } = useSyncExternalStore(adaptor.getHostContextStore("view").subscribe, adaptor.getHostContextStore("view").getSnapshot);
    const isOpen = mode === "modal";
    const handleBackdropClick = (e) => {
        if (e.target === e.currentTarget) {
            adaptor.closeModal();
        }
    };
    useEffect(() => {
        if (!isOpen) {
            return;
        }
        const handler = (e) => {
            if (e.key === "Escape") {
                adaptor.closeModal();
            }
        };
        document.addEventListener("keydown", handler);
        return () => document.removeEventListener("keydown", handler);
    }, [isOpen, adaptor]);
    return (_jsxs(_Fragment, { children: [_jsx("style", { children: modalStyles }), isOpen && (
            // biome-ignore lint/a11y/useKeyWithClickEvents: backdrop isn't focusable
            _jsx("div", { role: "dialog", className: "sb-modal-backdrop", onClick: handleBackdropClick })), _jsx("div", { className: isOpen ? "sb-modal-container" : undefined, children: children })] }));
}
