// skybridge/dist/web/bridges/apps-sdk/adaptor.js
import { AppsSdkBridge } from "./bridge.js";
export class AppsSdkAdaptor {
    static instance = null;
    static getInstance() {
        if (!AppsSdkAdaptor.instance) {
            AppsSdkAdaptor.instance = new AppsSdkAdaptor();
        }
        return AppsSdkAdaptor.instance;
    }
    static resetInstance() {
        AppsSdkAdaptor.instance = null;
    }
    getHostContextStore(key) {
        const bridge = AppsSdkBridge.getInstance();
        return {
            subscribe: bridge.subscribe(key),
            getSnapshot: () => bridge.getSnapshot(key),
        };
    }
    callTool = async (name, args) => {
        return window.openai.callTool(name, args);
    };
    requestDisplayMode = (mode) => {
        return window.openai.requestDisplayMode({ mode });
    };
    sendFollowUpMessage = (prompt) => {
        return window.openai.sendFollowUpMessage({ prompt });
    };
    openExternal(href) {
        window.openai.openExternal({ href });
    }
    setWidgetState = (stateOrUpdater) => {
        const newState = typeof stateOrUpdater === "function"
            ? stateOrUpdater(window.openai.widgetState)
            : stateOrUpdater;
        return window.openai.setWidgetState(newState);
    };
    uploadFile = (file) => {
        return window.openai.uploadFile(file);
    };
    getFileDownloadUrl = (file) => {
        return window.openai.getFileDownloadUrl(file);
    };
    openModal(options) {
        return window.openai.requestModal(options);
    }
    setOpenInAppUrl(href) {
        href = href.trim();
        if (!href) {
            throw new Error("The href parameter is required.");
        }
        const serverUrl = window.skybridge.serverUrl;
        if (!serverUrl) {
            throw new Error("The widgetDomain property has not been set on the widget resource meta object.");
        }
        const domainUrl = new URL(serverUrl);
        const hrefUrl = new URL(href, serverUrl);
        if (domainUrl.origin !== hrefUrl.origin) {
            throw new Error("Provided href is not compatible with widget domain: origin differs");
        }
        return window.openai.setOpenInAppUrl({ href });
    }
}
