// skybridge/dist/server/asset-base-url-transform-plugin.js
/**
 * Transforms asset import paths to use the development server base URL.
 */
export function assetBaseUrlTransform(code, devServerOrigin) {
    const assetStringPattern = /(?<!https?:\/\/)(["'`])(\/[^"'`]+\.(svg|png|jpeg|jpg|gif|webp|mp3|mp4|woff|woff2|ttf|eot))\1/g;
    code = code.replace(assetStringPattern, (_match, quote, assetPath) => {
        return `${quote}${devServerOrigin}${assetPath}${quote}`;
    });
    return code;
}
/**
 * Vite plugin that transforms asset import paths to use the development server base URL.
 */
export function assetBaseUrlTransformPlugin(options) {
    const { devServerOrigin } = options;
    return {
        name: "asset-base-url-transform",
        enforce: "pre",
        transform(code) {
            if (!code) {
                return null;
            }
            const transformedCode = assetBaseUrlTransform(code, devServerOrigin);
            if (transformedCode === code) {
                return null;
            }
            return {
                code: transformedCode,
                map: null,
            };
        },
    };
}
