// skybridge/dist/web/hooks/test/utils.js
import { fireEvent } from "@testing-library/react";
import { act } from "react";
import { vi } from "vitest";
export class MockResizeObserver {
    observe() { }
    unobserve() { }
    disconnect() { }
}
const DEFAULT_CONTEXT = {};
export const getMcpAppHostPostMessageMock = (initialContext = DEFAULT_CONTEXT) => vi.fn((message) => {
    switch (message.method) {
        case "ui/initialize": {
            const result = {
                protocolVersion: "2025-06-18",
                hostInfo: { name: "test-host", version: "1.0.0" },
                hostCapabilities: {},
                hostContext: initialContext,
            };
            act(() => fireEvent(window, new MessageEvent("message", {
                data: {
                    jsonrpc: "2.0",
                    id: message.id,
                    result,
                },
            })));
            break;
        }
        case "ui/update-model-context": {
            act(() => fireEvent(window, new MessageEvent("message", {
                data: {
                    jsonrpc: "2.0",
                    id: message.id,
                    result: {},
                },
            })));
            break;
        }
    }
});
export const fireToolInputNotification = (args) => {
    fireEvent(window, new MessageEvent("message", {
        data: {
            jsonrpc: "2.0",
            method: "ui/notifications/tool-input",
            params: {
                arguments: args,
            },
        },
    }));
};
export const fireToolResultNotification = (params) => {
    fireEvent(window, new MessageEvent("message", {
        data: {
            jsonrpc: "2.0",
            method: "ui/notifications/tool-result",
            params,
        },
    }));
};
