// skybridge/dist/web/data-llm.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useEffect, useId, } from "react";
import { getAdaptor } from "./bridges/index.js";
export const WIDGET_CONTEXT_KEY = "__widget_context";
const nodes = new Map();
function setNode(node) {
    nodes.set(node.id, node);
    onChange();
}
function removeNode(id) {
    nodes.delete(id);
    onChange();
}
function onChange() {
    const description = getLLMDescriptionString();
    const adaptor = getAdaptor();
    adaptor.setWidgetState((prevState) => ({
        ...prevState,
        [WIDGET_CONTEXT_KEY]: description,
    }));
}
const ParentIdContext = createContext(null);
export function DataLLM({ content, children }) {
    const parentId = useContext(ParentIdContext);
    const id = useId();
    useEffect(() => {
        if (content) {
            setNode({
                id,
                parentId,
                content,
            });
        }
        else {
            removeNode(id);
        }
        return () => {
            removeNode(id);
        };
    }, [id, parentId, content]);
    return (_jsx(ParentIdContext.Provider, { value: id, children: children }));
}
function getLLMDescriptionString() {
    const byParent = new Map();
    for (const node of Array.from(nodes.values())) {
        const key = node.parentId ?? null;
        if (!byParent.has(key)) {
            byParent.set(key, []);
        }
        byParent.get(key)?.push(node);
    }
    for (const list of byParent.values()) {
        list.sort((a, b) => a.id.localeCompare(b.id));
    }
    const lines = [];
    function traverseTree(parentId, depth) {
        const children = byParent.get(parentId);
        if (!children) {
            return;
        }
        for (const child of children) {
            if (child.content?.trim()) {
                const indent = "  ".repeat(depth);
                lines.push(`${indent}- ${child.content.trim()}`);
            }
            traverseTree(child.id, depth + 1);
        }
    }
    traverseTree(null, 0);
    return lines.join("\n");
}
