// alpic/dist/api.js
import {} from "./types.js";
export class ApiError extends Error {
    statusCode;
    constructor(message, statusCode) {
        super(message);
        this.statusCode = statusCode;
        this.name = "ApiError";
        Object.setPrototypeOf(this, ApiError.prototype);
    }
}
function getApiBaseUrl() {
    return process.env.ALPIC_API_BASE_URL ?? "https://api.alpic.ai";
}
export function getFrontendBaseUrl() {
    return process.env.ALPIC_FRONTEND_BASE_URL ?? "https://app.alpic.ai";
}
const doFetch = async (path, options = {}) => {
    const res = await fetch(`${getApiBaseUrl()}${path}`, {
        ...options,
        headers: {
            Authorization: `Bearer ${process.env.ALPIC_API_KEY}`,
            ...options.headers,
        },
    });
    if (!res.ok) {
        try {
            const body = await res.json();
            throw new ApiError(`${res.status} ${res.statusText}: ${body.message}`, res.status);
        }
        catch (error) {
            if (error instanceof ApiError)
                throw error;
            throw new ApiError(res.statusText, res.status);
        }
    }
    return res.json();
};
export const api = {
    listProjects: async () => {
        return doFetch("/v1/projects", { method: "GET" });
    },
    getProject: async (projectId) => {
        return doFetch(`/v1/projects/${projectId}`, { method: "GET" });
    },
    createProject: async (input) => {
        return doFetch("/v1/projects", {
            method: "POST",
            body: JSON.stringify(input),
            headers: { "Content-Type": "application/json" },
        });
    },
    getPresignedUploadUrl: async () => {
        return doFetch("/v1/deployments/upload", { method: "POST" });
    },
    deployEnvironment: async (environmentId, body) => {
        return doFetch(`/v1/environments/${environmentId}/deploy`, {
            method: "POST",
            body: JSON.stringify(body ?? {}),
            headers: { "Content-Type": "application/json" },
        });
    },
    getDeployment: async (deploymentId) => {
        return doFetch(`/v1/deployments/${deploymentId}`, { method: "GET" });
    },
    getEnvironment: async (environmentId) => {
        return doFetch(`/v1/environments/${environmentId}`, { method: "GET" });
    },
};
