// alpic/dist/lib/archive.js
import { execSync } from "node:child_process";
import { existsSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { create as tarCreate } from "tar";
const GIT_FILES_MAX_BUFFER = 10 * 1024 * 1024;
function isGitRepository(dir) {
    return existsSync(join(resolve(dir), ".git"));
}
export function ensureGitAvailable() {
    try {
        execSync("git --version", { stdio: "ignore" });
    }
    catch {
        throw new Error("Git is required to deploy. Please install git and ensure it is available in your PATH.");
    }
}
export function getGitFiles(deployDir) {
    const dir = resolve(deployDir);
    const output = execSync("git ls-files -z --cached --others --exclude-standard -- .", {
        cwd: dir,
        encoding: "utf8",
        maxBuffer: GIT_FILES_MAX_BUFFER,
    });
    const files = output.split("\0").filter(Boolean);
    if (files.length === 0) {
        throw new Error("No tracked or untracked files found. Ensure you are in a git repository with files to deploy.");
    }
    return files;
}
export function getFilesToPack(deployDir) {
    const dir = resolve(deployDir);
    if (isGitRepository(dir)) {
        return getGitFiles(deployDir);
    }
    ensureGitAvailable();
    execSync("git init", { cwd: dir });
    try {
        return getGitFiles(deployDir);
    }
    finally {
        rmSync(join(dir, ".git"), { recursive: true });
    }
}
export async function createTarArchive(files, deployDir) {
    const tmpDir = mkdtempSync(join(tmpdir(), "alpic-deploy-"));
    const archivePath = join(tmpDir, "source.tar.gz");
    await tarCreate({
        gzip: true,
        file: archivePath,
        cwd: deployDir,
    }, files);
    return { tmpDir, archivePath };
}
