// alpic/dist/lib/config.js
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
const getProjectJsonPath = (deployDir) => join(deployDir, ".alpic", "project.json");
export const config = {
    load: (deployDir) => {
        const path = getProjectJsonPath(deployDir);
        if (!existsSync(path)) {
            return null;
        }
        let raw;
        try {
            raw = JSON.parse(readFileSync(path, "utf8"));
        }
        catch (error) {
            throw new Error(`Failed to load project config: ${error.message}`);
        }
        return {
            projectId: raw.projectId,
            teamId: raw.teamId,
            projectName: raw.projectName ?? raw.projectId,
            environmentId: raw.environmentId,
            environmentName: raw.environmentName,
        };
    },
    save: (config, deployDir) => {
        const path = getProjectJsonPath(deployDir);
        mkdirSync(join(deployDir, ".alpic"), { recursive: true });
        writeFileSync(path, JSON.stringify(config, null, 2));
    },
};
