// alpic/dist/lib/global-config.js
import crypto from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
const GLOBAL_CONFIG_DIR = join(homedir(), ".alpic");
export const GLOBAL_CONFIG_FILE = join(GLOBAL_CONFIG_DIR, "config.json");
function readJsonFile(filePath) {
    try {
        if (existsSync(filePath)) {
            const content = readFileSync(filePath, "utf-8");
            return JSON.parse(content);
        }
    }
    catch {
        // Ignore errors reading config
    }
    return null;
}
function writeJsonFile(filePath, data) {
    try {
        const dir = join(filePath, "..");
        if (!existsSync(dir)) {
            mkdirSync(dir, { recursive: true });
        }
        writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
    }
    catch {
        // Ignore errors writing config
    }
}
export function getGlobalConfig() {
    const existing = readJsonFile(GLOBAL_CONFIG_FILE);
    if (existing?.machineId && existing?.telemetry !== undefined) {
        return existing;
    }
    const config = {
        machineId: existing?.machineId ?? crypto.randomUUID(),
        telemetry: {
            enabled: existing?.telemetry?.enabled ?? true,
        },
    };
    writeJsonFile(GLOBAL_CONFIG_FILE, config);
    return config;
}
export function saveGlobalConfig(config) {
    writeJsonFile(GLOBAL_CONFIG_FILE, config);
}
