// alpic/dist/lib/telemetry.js
import ci from "ci-info";
import crypto from "node:crypto";
import posthog from "../posthog.js";
import { getGlobalConfig, saveGlobalConfig } from "./global-config.js";
const ENV_TELEMETRY_DISABLED = "ALPIC_TELEMETRY_DISABLED";
const ENV_TELEMETRY_DEBUG = "ALPIC_TELEMETRY_DEBUG";
const ENV_DO_NOT_TRACK = "DO_NOT_TRACK";
export function isEnabled() {
    if (process.env[ENV_TELEMETRY_DISABLED] === "1" || process.env[ENV_TELEMETRY_DISABLED]?.toLowerCase() === "true") {
        return false;
    }
    if (process.env[ENV_DO_NOT_TRACK] === "1" || process.env[ENV_DO_NOT_TRACK]?.toLowerCase() === "true") {
        return false;
    }
    if (ci.isCI) {
        return true;
    }
    const config = getGlobalConfig();
    return config.telemetry.enabled;
}
export function isDebugMode() {
    return process.env[ENV_TELEMETRY_DEBUG] === "1" || process.env[ENV_TELEMETRY_DEBUG]?.toLowerCase() === "true";
}
export function setEnabled(enabled) {
    const config = getGlobalConfig();
    config.telemetry.enabled = enabled;
    saveGlobalConfig(config);
}
export function getMachineId() {
    if (ci.isCI) {
        return ci.name ?? "unknown-ci";
    }
    return getGlobalConfig().machineId;
}
const hook = async ({ id: command, config: { version }, error, }) => {
    if (!isEnabled()) {
        return;
    }
    const event = {
        command,
        version,
        machineId: getMachineId(),
        sessionId: crypto.randomUUID(),
        isCI: ci.isCI,
        nodeVersion: process.version,
        platform: process.platform,
        outcome: error ? "failure" : "success",
        error: error?.message,
    };
    if (isDebugMode()) {
        console.error("[Telemetry Debug] Would send event:", JSON.stringify(event, null, 2));
        return;
    }
    try {
        posthog.capture({
            distinctId: event.machineId,
            event: command,
            properties: event,
        });
    }
    catch (e) {
        // Silently ignore telemetry errors - never block CLI operation
    }
};
export default hook;
